### PR TITLE
Kotlin/Native compiler and Gradle plugins

### PR DESCRIPTION
**Pull request by charleskorn**
See #555.

This includes a temporary workaround for https://youtrack.jetbrains.com/issue/KT-29901.

### CODE REVIEW

**Comment by raniejade** [comment]
Oh wow, this awesome @charleskorn! 💯 . Going to take a look next week, I'll be on holidays this week.

**Comment by raniejade** [comment]
@charleskorn sorry for not getting back on you yet, it's a very busy week at work. Hopefully, I can have a bit of free time on the weekends or next week.

**Comment by charleskorn** [comment]
No worries, there's no rush!

**Comment by raniejade** [comment]
I had a quick look and it's looking good. I'll have to test it out locally first but I'll merge this soon. The rough plan is to publish `2.0.2` first then merge this. From `2.0.3` it will be available but is experimental - I'll work on fixing the IDE plugin to support KN. Good job @charleskorn! 💯

**Comment by charleskorn** [comment]
Awesome, thanks!

**Comment by raniejade** [comment]
@charleskorn sorry, delaying the merge a bit to `2.0.4`, there were some critical bugs in `2.0.2` that needs to go out first.

File: spek-kotlin-native-compiler-plugin/src/main/kotlin/org/spekframework/spek2/kotlincompilerplugin/SpekExtension.kt [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+import org.jetbrains.kotlin.ir.visitors.acceptChildrenVoid
+import org.jetbrains.kotlin.name.ClassId
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+
+class SpekExtension : IrGenerationExtension {
+    override fun generate(file: IrFile, backendContext: BackendContext, bindingContext: BindingContext) {
+        val spekCollector = SpekCollector(file, backendContext)
```
Comment (raniejade) [inline spek-kotlin-native-compiler-plugin/src/main/kotlin/org/spekframework/spek2/kotlincompilerplugin/SpekExtension.kt:52]: @charleskorn just started the `2.0.3` release, looking at this for the next release. For context, what is `file` here?

Comment (charleskorn) [reply spek-kotlin-native-compiler-plugin/src/main/kotlin/org/spekframework/spek2/kotlincompilerplugin/SpekExtension.kt:52]: `file` is a Kotlin source code file.

Comment (raniejade) [reply spek-kotlin-native-compiler-plugin/src/main/kotlin/org/spekframework/spek2/kotlincompilerplugin/SpekExtension.kt:52]: which file exactly? Is it specified somewhere?

Comment (charleskorn) [reply spek-kotlin-native-compiler-plugin/src/main/kotlin/org/spekframework/spek2/kotlincompilerplugin/SpekExtension.kt:52]: I'm unclear what your question is here sorry - this method gets called for every Kotlin source file in the current compilation operation.

Comment (raniejade) [reply spek-kotlin-native-compiler-plugin/src/main/kotlin/org/spekframework/spek2/kotlincompilerplugin/SpekExtension.kt:52]: > this method gets called for every Kotlin source file in the current compilation operation.

Ahh, right. Was missing this, thanks for the clarification!

Comment (charleskorn) [reply spek-kotlin-native-compiler-plugin/src/main/kotlin/org/spekframework/spek2/kotlincompilerplugin/SpekExtension.kt:52]: No worries!

**Comment by raniejade** [comment]
Merged at 0eeb7697efc9429e9da7aa78ceecf87e2a652261, had to fix some conflicts locally. Thanks heaps @charleskorn!

**Pull request closed without merging**